File: android/src/main/kotlin/twilio/flutter/twilio_programmable_chat/methods/MessageMethods.kt
package twilio.flutter.twilio_programmable_chat.methods

import com.twilio.chat.CallbackListener
import com.twilio.chat.Channel
import com.twilio.chat.ErrorInfo
import com.twilio.chat.Message
import com.twilio.chat.ProgressListener
import com.twilio.chat.StatusListener
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.File
import org.json.JSONException
import twilio.flutter.twilio_programmable_chat.Mapper
import twilio.flutter.twilio_programmable_chat.TwilioProgrammableChatPlugin

object MessageMethods {
    fun getChannel(call: MethodCall, result: MethodChannel.Result) {
        val channelSid = call.argument<String>("channelSid")
                ?: return result.error("ERROR", "Missing 'channelSid'", null)

        TwilioProgrammableChatPlugin.chatListener.chatClient?.channels?.getChannel(channelSid, object : CallbackListener<Channel>() {
            override fun onSuccess(channel: Channel) {
                TwilioProgrammableChatPlugin.debug("MessageMethods.getChannel => onSuccess")
                result.success(Mapper.channelToMap(channel))
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioProgrammableChatPlugin.debug("MessageMethods.getChannel => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }

    fun updateMessageBody(call: MethodCall, result: MethodChannel.Result) {
        val channelSid = call.argument<String>("channelSid")
                ?: return result.error("ERROR", "Missing 'channelSid'", null)

        val messageIndex = call.argument<Int>("messageIndex")?.toLong()
                ?: return result.error("ERROR", "Missing 'messageIndex'", null)

        val body = call.argument<String>("body")
                ?: return result.error("ERROR", "Missing 'body'", null)

        TwilioProgrammableChatPlugin.chatListener.chatClient?.channels?.getChannel(channelSid, object : CallbackListener<Channel>() {
            override fun onSuccess(channel: Channel) {
                TwilioProgrammableChatPlugin.debug("MessageMethods.updateMessageBody (Channels.getChannel) => onSuccess")

                channel.messages.getMessageByIndex(messageIndex, object : CallbackListener<Message>() {
                    override fun onSuccess(message: Message) {
                        TwilioProgrammableChatPlugin.debug("MessageMethods.updateMessageBody (Messages.getMessageByIndex) => onSuccess")

                        message.updateMessageBody(body, object : StatusListener() {
                            override fun onSuccess() {
                                TwilioProgrammableChatPlugin.debug("MessageMethods.updateMessageBody (Message.updateMessageBody) => onSuccess")
                                result.success(message.messageBody)
                            }

                            override fun onError(errorInfo: ErrorInfo) {
                                TwilioProgrammableChatPlugin.debug("MessageMethods.updateMessageBody (Message.updateMessageBody) => onError: $errorInfo")
                                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                            }
                        })
                    }

                    override fun onError(errorInfo: ErrorInfo) {
                        TwilioProgrammableChatPlugin.debug("MessageMethods.updateMessageBody (Messages.getMessageByIndex) => onError: $errorInfo")
                        result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                    }
                })
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioProgrammableChatPlugin.debug("MessageMethods.updateMessageBody (Channels.getChannel) => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }

    fun setAttributes(call: MethodCall, result: MethodChannel.Result) {
        val channelSid = call.argument<String>("channelSid")
                ?: return result.error("ERROR", "Missing 'channelSid'", null)

        val messageIndex = call.argument<Int>("messageIndex")?.toLong()
                ?: return result.error("ERROR", "Missing 'messageIndex'", null)

        // Not erroring out because a nullable attributes is allowed to reset the Channel attributes.
        val attributes = call.argument<Map<String, Any>>("attributes")

        try {
            TwilioProgrammableChatPlugin.chatListener.chatClient?.channels?.getChannel(channelSid, object : CallbackListener<Channel>() {
                override fun onSuccess(channel: Channel) {
                    TwilioProgrammableChatPlugin.debug("MessageMethods.setAttributes => onSuccess")

                    channel.messages.getMessageByIndex(messageIndex, object : CallbackListener<Message>() {
                        override fun onSuccess(message: Message) {
                            TwilioProgrammableChatPlugin.debug("MessageMethods.updateMessageBody (Messages.getMessageByIndex) => onSuccess")
                            message.setAttributes(Mapper.mapToAttributes(attributes), object : StatusListener() {
                                override fun onSuccess() {
                                    TwilioProgrammableChatPlugin.debug("MessageMethods.setAttributes  (Channel.setAttributes) => onSuccess")
                                    try {
                                        result.success(Mapper.attributesToMap(message.attributes))
                                    } catch (err: JSONException) {
                                        return result.error("JSONException", err.message, null)
                                    }
                                }

                                override fun onError(errorInfo: ErrorInfo) {
                                    TwilioProgrammableChatPlugin.debug("MessageMethods.setAttributes  (Channel.setAttributes) => onError: $errorInfo")
                                    result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                                }
                            })
                            try {
                                result.success(Mapper.attributesToMap(message.attributes))
                            } catch (err: JSONException) {
                                return result.error("JSONException", err.message, null)
                            }
                        }

                        override fun onError(errorInfo: ErrorInfo) {
                            TwilioProgrammableChatPlugin.debug("MessageMethods.updateMessageBody (Messages.getMessageByIndex) => onError: $errorInfo")
                            result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                        }
                    })
                }

                override fun onError(errorInfo: ErrorInfo) {
                    TwilioProgrammableChatPlugin.debug("MessageMethods.setAttributes => onError: $errorInfo")
                    result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                }
            })
        } catch (err: IllegalArgumentException) {
            return result.error("IllegalArgumentException", err.message, null)
        }
    }

    fun getMedia(call: MethodCall, result: MethodChannel.Result) {
        val channelSid = call.argument<String>("channelSid")
                ?: return result.error("ERROR", "Missing 'channelSid'", null)

        val messageIndex = call.argument<Int>("messageIndex")?.toLong()
                ?: return result.error("ERROR", "Missing 'messageIndex'", null)

        val path = call.argument<String>("filePath")
                ?: return result.error("ERROR", "Missing 'filePath'", null)

        val file = File(path)
        if (file.exists() && file.length() > 0) {
            result.success(true)
        } else {
            TwilioProgrammableChatPlugin.debug("Downloading media for message $messageIndex to path: $path")
            TwilioProgrammableChatPlugin.chatListener.chatClient?.channels?.getChannel(channelSid, object : CallbackListener<Channel>() {
                override fun onSuccess(channel: Channel) {
                    TwilioProgrammableChatPlugin.debug("MessageMethods.getMedia => onSuccess")
                    channel.messages.getMessageByIndex(messageIndex, object : CallbackListener<Message>() {
                        override fun onSuccess(message: Message) {
                            TwilioProgrammableChatPlugin.debug("MessageMethods.getMedia (Messages.getMessageByIndex) => onSuccess")
                            message.media.download(file.outputStream(), object : StatusListener() {
                                override fun onSuccess() {
                                    TwilioProgrammableChatPlugin.debug("MessageMethods.getMedia (Message.Media.download) => onSuccess")
                                    result.success(true)
                                }

                                override fun onError(errorInfo: ErrorInfo) {
                                    TwilioProgrammableChatPlugin.debug("MessageMethods.getMedia (Message.Media.download) => onError: $errorInfo")
                                    result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                                }
                            }, object : ProgressListener() {
                                override fun onStarted() {
                                    TwilioProgrammableChatPlugin.debug("MessageMethods.getMedia (Message.Media.download) => onStarted")
                                }

                                override fun onProgress(bytes: Long) {
                                    TwilioProgrammableChatPlugin.debug("MessageMethods.getMedia (Message.Media.download) => onProgress: $bytes")
                                }

                                override fun onCompleted(mediaSid: String?) {
                                    TwilioProgrammableChatPlugin.debug("MessageMethods.getMedia (Message.Media.download) => onCompleted: $mediaSid")
                                }
                            })
                        }

                        override fun onError(errorInfo: ErrorInfo) {
                            TwilioProgrammableChatPlugin.debug("MessageMethods.updateMessageBody (Messages.getMessageByIndex) => onError: $errorInfo")
                            result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                        }
                    })
                }

                override fun onError(errorInfo: ErrorInfo) {
                    TwilioProgrammableChatPlugin.debug("MessageMethods.getMedia => onError: $errorInfo")
                    result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                }
            })
        }
    }
}

File: ios/Classes/Methods/MessageMethods.swift
import Flutter
import Foundation
import TwilioChatClient

public class MessageMethods {
    public static func updateMessageBody(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        guard let channelSid = arguments["channelSid"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'channelSid' parameter", details: nil))
        }

        guard let messageIndex = arguments["messageIndex"] as? NSNumber else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'messageIndex' parameter", details: nil))
        }

        guard let body = arguments["body"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'body' parameter", details: nil))
        }

        let flutterResult = result
        SwiftTwilioProgrammableChatPlugin.chatListener?.chatClient?.channelsList()?.channel(withSidOrUniqueName: channelSid, completion: { (result: TCHResult, channel: TCHChannel?) in
            if result.isSuccessful(), let channel = channel {
                SwiftTwilioProgrammableChatPlugin.debug("MessageMethods.updateMessageBody (Channels.getChannel) => onSuccess")
                channel.messages?.message(withIndex: messageIndex, completion: { (result: TCHResult, message: TCHMessage?) in
                    if result.isSuccessful(), let message = message {
                        SwiftTwilioProgrammableChatPlugin.debug("MessageMethods.updateMessageBody (Messages.messageWithIndex) => onSuccess")
                        message.updateBody(body) { (result: TCHResult) in
                            if result.isSuccessful() {
                                SwiftTwilioProgrammableChatPlugin.debug("MessageMethods.updateMessageBody (Message.updateMessageBody) => onSuccess")
                                flutterResult(message.body)
                            }
                        }
                    } else {
                        SwiftTwilioProgrammableChatPlugin.debug("MessageMethods.updateMessageBody (Messages.messageWithIndex) => onError: \(String(describing: result.error))")
                        flutterResult(FlutterError(code: "ERROR", message: "Error retrieving message (index: \(messageIndex)) from channel (sid: \(channelSid)", details: nil))
                    }
                })
            } else {
                if let error = result.error as NSError? {
                    SwiftTwilioProgrammableChatPlugin.debug("MessageMethods.updateMessageBody => onError: \(error)")
                    flutterResult(FlutterError(code: "ERROR", message: "Error retrieving channel (sid: \(channelSid))", details: nil))
                }
            }
        })
    }

    public static func setAttributes(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        guard let channelSid = arguments["channelSid"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'channelSid' parameter", details: nil))
        }

        guard let messageIndex = arguments["messageIndex"] as? NSNumber else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'messageIndex' parameter", details: nil))
        }

        guard let attributesDict = arguments["attributes"] as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'attributes' parameter", details: nil))
        }

        let attributes = TCHJsonAttributes.init(dictionary: attributesDict as [AnyHashable: Any])

        let flutterResult = result
        SwiftTwilioProgrammableChatPlugin.chatListener?.chatClient?.channelsList()?.channel(withSidOrUniqueName: channelSid, completion: { (result: TCHResult, channel: TCHChannel?) in
            if result.isSuccessful(), let channel = channel {
                SwiftTwilioProgrammableChatPlugin.debug("MessageMethods.setAttributes => onSuccess")
                channel.messages?.message(withIndex: messageIndex, completion: { (result: TCHResult, message: TCHMessage?) in
                    if result.isSuccessful(), let message = message {
                        SwiftTwilioProgrammableChatPlugin.debug("MessageMethods.setAttributes (Messages.messageWithIndex) => onSuccess")
                        message.setAttributes(attributes) { (result: TCHResult) in
                            if result.isSuccessful() {
                                SwiftTwilioProgrammableChatPlugin.debug("MessageMethods.setAttributes (Message.setAttributes) => onSuccess")
                                flutterResult(Mapper.attributesToDict(message.attributes()))
                            } else {
                                SwiftTwilioProgrammableChatPlugin.debug("MessageMethods.setAttributes (Message.setAttributes) => onError: \(String(describing: result.error))")
                            }
                        }
                    } else {
                        SwiftTwilioProgrammableChatPlugin.debug("MessageMethods.setAttributes (Messages.messageWithIndex) => onError: \(String(describing: result.error))")
                        flutterResult(FlutterError(code: "ERROR", message: "Error retrieving message (index: \(messageIndex)) from channel (sid: \(channelSid)", details: nil))
                    }
                })
            } else {
                if let error = result.error as NSError? {
                    SwiftTwilioProgrammableChatPlugin.debug("MessageMethods.setAttributes => onError: \(error)")
                    flutterResult(FlutterError(code: "ERROR", message: "Error retrieving channel with sid or uniqueName '\(channelSid)'", details: nil))
                }
            }
        })
    }

    public static func getMedia(_ call: FlutterMethodCall, result flutterResult: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?],
            let channelSid = arguments["channelSid"] as? String,
            let messageIndex = arguments["messageIndex"] as? NSNumber,
            let path = arguments["filePath"] as? String else {
                return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        let manager = FileManager.default

        var fileSize: UInt64
        do {
            let attr = try manager.attributesOfItem(atPath: path) as NSDictionary
            fileSize = attr.fileSize()
        } catch {
            fileSize = 0
        }

        if manager.fileExists(atPath: path), fileSize > 0 {
            flutterResult(true)
        } else {
            SwiftTwilioProgrammableChatPlugin.chatListener?.chatClient?.channelsList()?.channel(withSidOrUniqueName: channelSid, completion: { (result: TCHResult, channel: TCHChannel?) in
                if result.isSuccessful(), let channel = channel {
                    channel.messages?.message(withIndex: messageIndex, completion: { (result: TCHResult, message: TCHMessage?) in
                        if result.isSuccessful(), let message = message {
                            let url = URL(fileURLWithPath: path)
                            if let outputStream = OutputStream(url: url, append: false) {
                                message.getMediaWith(outputStream, onStarted: {
                                    SwiftTwilioProgrammableChatPlugin.debug("getMediaWith => onStarted")
                                }, onProgress: { (bytes: UInt) in
                                    SwiftTwilioProgrammableChatPlugin.debug("getMediaWith => onProgress: \(bytes)")
                                }, onCompleted: { (mediaSid: String) in
                                    SwiftTwilioProgrammableChatPlugin.debug("getMediaWith => onCompleted: \(mediaSid)")
                                }, completion: { (result: TCHResult) in
                                    SwiftTwilioProgrammableChatPlugin.debug("getMediaWith => completion: \(result.isSuccessful())")
                                    flutterResult(result.isSuccessful())
                                })
                            } else {
                                flutterResult(FlutterError(code: "ERROR", message: "Error opening file at path '\(path)'", details: nil))
                            }
                        } else {
                            flutterResult(FlutterError(code: "ERROR", message: "Error retrieving message with index '\(messageIndex)'", details: nil))
                        }
                    })
                } else {
                    flutterResult(FlutterError(code: "ERROR", message: "Error retrieving channel with sid '\(channelSid)'", details: nil))
                }
            })
        }
    }
}
